Implement terraform based system tests

mtojek opened the pull request:
~I'm opening this as draft to present the progress on the TF-based system tests.~

~Please do not consider this as final form, as I will modify it soon.~

Issue: https://github.com/elastic/elastic-package/issues/89

How to test this:

```
export AWS_ACCESS_KEY_ID="secret"
export AWS_SECRET_ACCESS_KEY="secret"

cd test/packages/aws
elastic-package test system --data-streams ec2_metrics -v
```

Tf definitions in: `data_stream/ec2_metrics/_dev/deploy/tf/main.tf` (or other .tf files)

EDIT:

I switched from SNS to EC2 metrics during tests as metrics tend to appear quicker.

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-227/detail/PR-227/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-227/detail/PR-227/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-227/detail/PR-227/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-227/detail/PR-227/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-package_227.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Branch indexing`



* * Start Time: 2021-01-19T22:28:26.149+0000


* Duration: 2 min 28 sec


* Commit: fe54067ad860e2b817eada19831f88aaf265dc79




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->





### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package%2FPR-227/detail/PR-227/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
##### `Check`
<ul>
<li>Took 1 min 30 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/elastic-package/pipelines/PR-227/runs/1/steps/123/log/?start=0">here</a></li>
<li>Description: <code>make check</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-01-19T22:30:23.966Z] go: downloading github.com/jbenet/go-context v0.0.0-20150711004518-d14ea06fba99
[2021-01-19T22:30:23.966Z] go: downloading gopkg.in/warnings.v0 v0.1.2
[2021-01-19T22:30:24.226Z] go: downloading github.com/mattn/go-runewidth v0.0.9
[2021-01-19T22:30:24.227Z] go: downloading github.com/go-openapi/strfmt v0.19.6
[2021-01-19T22:30:24.227Z] go: downloading go.mongodb.org/mongo-driver v1.0.3
[2021-01-19T22:30:24.227Z] go: downloading github.com/go-openapi/errors v0.19.2
[2021-01-19T22:30:24.227Z] go: downloading github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
[2021-01-19T22:30:24.227Z] go: downloading github.com/mitchellh/mapstructure v1.1.2
[2021-01-19T22:30:24.490Z] go: downloading github.com/go-stack/stack v1.8.0
[2021-01-19T22:30:39.392Z] go get -u golang.org/x/tools/cmd/goimports
[2021-01-19T22:30:39.392Z] go: downloading golang.org/x/tools v0.1.0
[2021-01-19T22:30:39.392Z] go: found golang.org/x/tools/cmd/goimports in golang.org/x/tools v0.1.0
[2021-01-19T22:30:39.392Z] go: downloading golang.org/x/tools v0.0.0-20210108195828-e2f9c7f1fc8e
[2021-01-19T22:30:39.392Z] go: downloading golang.org/x/mod v0.3.0
[2021-01-19T22:30:39.392Z] go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
[2021-01-19T22:30:39.392Z] go: golang.org/x/tools upgrade => v0.1.0
[2021-01-19T22:30:39.392Z] go: golang.org/x/mod upgrade => v0.4.1
[2021-01-19T22:30:39.392Z] go: golang.org/x/xerrors upgrade => v0.0.0-20200804184101-5ec99f83aff1
[2021-01-19T22:30:39.392Z] go: downloading golang.org/x/sys v0.0.0-20210119212857-b64e53b001e4
[2021-01-19T22:30:39.392Z] go: downloading golang.org/x/mod v0.4.1
[2021-01-19T22:30:39.392Z] go: golang.org/x/sys upgrade => v0.0.0-20210119212857-b64e53b001e4
[2021-01-19T22:30:40.330Z] goimports -local github.com/elastic/elastic-package/ -w .
[2021-01-19T22:30:40.590Z] go get -u golang.org/x/lint/golint
[2021-01-19T22:30:40.590Z] go: downloading golang.org/x/lint v0.0.0-20201208152925-83fdc39ff7b5
[2021-01-19T22:30:40.590Z] go: found golang.org/x/lint/golint in golang.org/x/lint v0.0.0-20201208152925-83fdc39ff7b5
[2021-01-19T22:30:40.590Z] go: golang.org/x/tools upgrade => v0.1.0
[2021-01-19T22:30:41.530Z] go list ./... | xargs -n 1 golint -set_exit_status
[2021-01-19T22:30:46.805Z] go get -u github.com/elastic/go-licenser
[2021-01-19T22:30:46.805Z] go: downloading github.com/elastic/go-licenser v0.3.1
[2021-01-19T22:30:46.805Z] go: github.com/elastic/go-licenser upgrade => v0.3.1
[2021-01-19T22:30:46.805Z] go-licenser -license Elastic
[2021-01-19T22:30:46.805Z] go mod tidy
[2021-01-19T22:30:46.805Z] go: downloading github.com/stretchr/testify v1.6.1
[2021-01-19T22:30:46.805Z] go: downloading gopkg.in/check.v1 v1.0.0-20200227125254-8fa46927fb4f
[2021-01-19T22:30:46.805Z] go: downloading github.com/hinshun/vt10x v0.0.0-20180616224451-1954e6464174
[2021-01-19T22:30:46.805Z] go: downloading github.com/go-git/go-git-fixtures/v4 v4.0.1
[2021-01-19T22:30:46.805Z] go: downloading github.com/google/uuid v1.1.1
[2021-01-19T22:30:46.805Z] go: downloading github.com/Netflix/go-expect v0.0.0-20180615182759-c93bf25de8e8
[2021-01-19T22:30:46.805Z] go: downloading github.com/google/go-cmp v0.5.1
[2021-01-19T22:30:46.805Z] go: downloading github.com/tidwall/pretty v1.0.0
[2021-01-19T22:30:46.805Z] go: downloading github.com/kr/pty v1.1.4
[2021-01-19T22:30:46.805Z] go: downloading github.com/armon/go-socks5 v0.0.0-20160902184237-e75332964ef5
[2021-01-19T22:30:46.805Z] go: downloading github.com/niemeyer/pretty v0.0.0-20200227124842-a10e7caefd8e
[2021-01-19T22:30:46.805Z] go: downloading github.com/gliderlabs/ssh v0.2.2
[2021-01-19T22:30:46.805Z] go: downloading github.com/davecgh/go-spew v1.1.1
[2021-01-19T22:30:46.805Z] go: downloading github.com/pmezard/go-difflib v1.0.0
[2021-01-19T22:30:46.805Z] go: downloading github.com/kr/text v0.2.0
[2021-01-19T22:30:46.805Z] go: downloading github.com/anmitsu/go-shlex v0.0.0-20161002113705-648efa622239
[2021-01-19T22:30:46.805Z] go: downloading github.com/flynn/go-shlex v0.0.0-20150515145356-3f9db97f8568
[2021-01-19T22:30:52.091Z] go: downloading github.com/alcortesm/tgz v0.0.0-20161220082320-9c5fe88206d7
[2021-01-19T22:30:52.358Z] git update-index --really-refresh
[2021-01-19T22:30:52.358Z] go.mod: needs update
[2021-01-19T22:30:52.358Z] go.sum: needs update
[2021-01-19T22:30:52.358Z] Makefile:40: recipe for target 'check-git-clean' failed
[2021-01-19T22:30:52.358Z] make: *** [check-git-clean] Error 1
[2021-01-19T22:30:52.433Z] Post stage
[2021-01-19T22:30:52.452Z] Running in /var/lib/jenkins/workspace/t-manager_elastic-package_PR-227/src/github.com/elastic/elastic-package
[2021-01-19T22:30:52.483Z] Archiving artifacts
[2021-01-19T22:30:52.508Z] Archiving artifacts
[2021-01-19T22:30:52.530Z] Recording test results
[2021-01-19T22:30:52.620Z] No test report files were found. Configuration error?
[2021-01-19T22:30:52.653Z] Error when executing always post condition:
[2021-01-19T22:30:52.658Z] Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from beats-ci-immutable-ubuntu-1804-1611095311975679186.c.elastic-ci-prod.internal/10.224.0.141:42446
[2021-01-19T22:30:52.658Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
[2021-01-19T22:30:52.658Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:357)
[2021-01-19T22:30:52.658Z] 		at hudson.remoting.Channel.call(Channel.java:1001)
[2021-01-19T22:30:52.658Z] 		at hudson.FilePath.act(FilePath.java:1070)
[2021-01-19T22:30:52.658Z] 		at hudson.FilePath.act(FilePath.java:1059)
[2021-01-19T22:30:52.658Z] 		at hudson.tasks.junit.JUnitParser.parseResult(JUnitParser.java:107)
[2021-01-19T22:30:52.658Z] 		at hudson.tasks.junit.JUnitResultArchiver.parse(JUnitResultArchiver.java:149)
[2021-01-19T22:30:52.658Z] 		at hudson.tasks.junit.JUnitResultArchiver.parseAndSummarize(JUnitResultArchiver.java:243)
[2021-01-19T22:30:52.658Z] 		at hudson.tasks.junit.pipeline.JUnitResultsStepExecution.run(JUnitResultsStepExecution.java:52)
[2021-01-19T22:30:52.658Z] 		at hudson.tasks.junit.pipeline.JUnitResultsStepExecution.run(JUnitResultsStepExecution.java:25)
[2021-01-19T22:30:52.658Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2021-01-19T22:30:52.658Z] 		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2021-01-19T22:30:52.658Z] 		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2021-01-19T22:30:52.658Z] 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-01-19T22:30:52.658Z] 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-01-19T22:30:52.658Z] hudson.AbortException: No test report files were found. Configuration error?
[2021-01-19T22:30:52.658Z] 	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:152)
[2021-01-19T22:30:52.658Z] 	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3112)
[2021-01-19T22:30:52.658Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
[2021-01-19T22:30:52.658Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2021-01-19T22:30:52.658Z] 	at hudson.remoting.Request$2.run(Request.java:375)
[2021-01-19T22:30:52.658Z] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:73)
[2021-01-19T22:30:52.658Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2021-01-19T22:30:52.658Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-01-19T22:30:52.658Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-01-19T22:30:52.658Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:118)
[2021-01-19T22:30:52.658Z] 	at java.lang.Thread.run(Thread.java:748)
[2021-01-19T22:30:52.658Z] 
[2021-01-19T22:30:53.057Z] Running on Jenkins in /var/lib/jenkins/workspace/t-manager_elastic-package_PR-227
[2021-01-19T22:30:53.268Z] [INFO] getVaultSecret: Getting secrets
[2021-01-19T22:30:53.348Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-01-19T22:30:54.053Z] + chmod 755 generate-build-data.sh
[2021-01-19T22:30:54.053Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-227/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-227/runs/1 FAILURE 147638
[2021-01-19T22:30:54.303Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-227/runs/1/steps/?limit=10000 -o steps-info.json
[2021-01-19T22:30:55.214Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Ingest-manager/elastic-package/PR-227/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-01-19T22:30:55.214Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-01-19T22:30:56.125Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mtojek commented:
Status update:

I managed to write the TF service deployer, but it needs to nice Terraform definition, which will actually perform something useful. A test scenario consist of two parts:
1. Test configuration (mainly vars): https://github.com/mtojek/elastic-package/blob/89-terraform-definitions/test/packages/aws/data_stream/sns/_dev/test/system/test-default-config.yml
2. Terraform definition which will bring up required AWS resources: https://github.com/mtojek/elastic-package/blob/89-terraform-definitions/test/packages/aws/data_stream/sns/_dev/deploy/tf/main.tf

@kaiyan-sheng I was wondering if you wouldn't like to play with it, share some feedback and write some Terraform files which can usually do sth useful (most wanted). I wrote a sample which creates SNS topic, but due to lack of messages, I don't see any CloudWatch metrics :(

FYI I haven't enabled the AWS account in the CI due to security reasons and cost-savings. Let's first check it on our workstations.

kaiyan-sheng commented:
@mtojek Will do(next week)! Thank you! I think if we can make the sns working, we can merge this and then add more tests to cover more datasets in other PRs.

mtojek commented:
@ycombinator With precious help of @kaiyan-sheng I managed to successfully follow the entire flow. This PR is ready for review now, but it depends on https://github.com/elastic/elastic-package/pull/228 .

kaiyan-sheng commented on test/packages/aws/docs/README.md:
```diff
@@ -1311,10 +1311,14 @@ An example event for `ec2` looks as following:
 | aws.ec2.cpu.total.pct | The percentage of allocated EC2 compute units that are currently in use on the instance. | scaled_float |
 | aws.ec2.diskio.read.bytes | Bytes read from all instance store volumes available to the instance. | long |
 | aws.ec2.diskio.read.bytes_per_sec | Bytes read per second from all instance store volumes available to the instance. | long |
+| aws.ec2.diskio.read.count | The number of disk IO reads | long |
```
Thanks for adding the fields!!!

kaiyan-sheng replied:
Oh I just realized this is added in `elastic-package`. How can we sync `elastic/integrations/packages/aws` with `elastic-package/test.packages/aws`?

mtojek replied:
The goal of these packages in the elastic-package repo is to have a playground to test new feature before they reach out to Integrations. Syncing is actually the other way round - once we apply similar changes in the Integrations, I will copy over it from Integrations to the elastic-package :)

kaiyan-sheng replied:
👍 Sounds good!

kaiyan-sheng commented on test/packages/aws/data_stream/ec2_metrics/_dev/test/system/test-default-config.yml:
```diff
@@ -0,0 +1,9 @@
+vars:
+  access_key_id: '{{AWS_ACCESS_KEY_ID}}'
+  secret_access_key: '{{AWS_SECRET_ACCESS_KEY}}'
```
Can we also add session_token in vars?

mtojek replied:
Done, added support for session token.

ycombinator commented on internal/install/install.go:
```diff
@@ -89,6 +98,21 @@ func ServiceLogsDir() (string, error) {
 }
 
+// ServiceDeployerComposeFile function returns the path to the selected service deployer's Docker Compose definition.
+func ServiceDeployerComposeFile(deployerName string) (string, error) {
```
Nit: I think you can just call this `TerraformDeployerComposeFile` and remove the `deployerName` argument. If/when we have more deployers that use a docker compose file for setup, we can generalize this then.

mtojek replied:
Fixed

ycombinator commented on internal/install/install.go:
```diff
@@ -139,6 +163,22 @@ func writeStackResources(elasticPackagePath string) error {
 	return nil
 }
 
+func writeDeployerResources(elasticPackagePath string) error {
```
Nit: since this is specific to terraform (for now at least), consider renaming it to `writeTerraformDeployerResources`. We can generalize this later if/when we have another deployer that needs resources to be written.

mtojek replied:
Fixed

ycombinator commented on internal/install/static_terraform_deployer_yml.go:
```diff
@@ -0,0 +1,20 @@
+    build: .
+    tty: true
+    environment:
+      - AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID} # TODO Extract to the env-file
```
Do you plan to address this TODO in this PR or a follow up PR? If it's for a follow up PR, can you make an issue for it so we don't lose track of it and post the link to the issue in the comment?

mtojek replied:
https://github.com/elastic/elastic-package/issues/235

ycombinator commented on internal/testrunner/runners/system/runner.go:
```diff
@@ -174,6 +176,10 @@ func (r *runner) run() (results []testrunner.TestResult, err error) {
 }
 
+func createTestRunID() string {
+	return fmt.Sprintf("%d", rand.Intn(99999-10000)+10000)
```
Maybe seed the PRNG each time with something like `rand.Seed(time.Now().UnixNano())`?

Also, nit: for an extra bit of readability, consider defining min and max constants and then using those in the `rand.Intn` call.

mtojek replied:
Fixed

ycombinator commented on internal/testrunner/runners/system/servicedeployer/factory.go:
```diff
@@ -5,27 +5,66 @@
+func Factory(options FactoryOptions) (ServiceDeployer, error) {
+	devDeployPath, err := findDevDeployPath(options)
+	if err != nil {
+		logger.Errorf("can't find _dev/deploy directory")
```
You're defining `devDeployDir` as a constant earlier. Might as well use it here. 🙂

mtojek replied:
Fixed

ycombinator commented on internal/testrunner/runners/system/servicedeployer/factory.go:
```diff
@@ -5,27 +5,66 @@
+	if _, err := os.Stat(terraformDirPath); err == nil {
+		return NewTerraformServiceDeployer(terraformDirPath)
+	}
 	return nil, ErrNotFound
```
Nit: please add newline to separate the final return statement from the two deployer cases.

mtojek replied:
Added, but I'm wondering why is this preference here :) ?

ycombinator replied:
Total personal preference, I suppose :) To me the two cases seem to be their own separate logical blocks from everything around them. But this is one of those things you can totally ignore in the future. In general, feel free to ignore anything I prefix with `Nit: `.

ycombinator commented on internal/testrunner/runners/system/servicedeployer/factory.go:
```diff
@@ -5,27 +5,66 @@
 	}
 
+	// Is the service defined using Terraform definition files?
+	terraformDirPath := filepath.Join(devDeployPath, "tf")
```
With the addition of the terraform deployer and the logic in this `Factory` function, we now have a precedence ordering. If a data set somehow defines folders for both types of deployers we will choose the docker compose one. Maybe this is something we should document in the system testing HOWTO document?

mtojek replied:
I described the change in docs (in this PR) and also described a bit the terraform service deployer (but without diving into details).

ycombinator replied:
Hmm, maybe I missed it but I'm not seeing anything in the docs in this PR that talk about the precedence ordering between the service deployers. I'm looking for something in the docs that would explain that only one service deployer should be defined for a package and if multiple ones are defined the Docker Compose service deployer will take precedence (or we can say the results are unpredictable).

ycombinator replied:
How about also adding a check in this code to see if more than one service deployer is defined and then return an error in that case?

mtojek replied:
I admit I forgot about his idea. Let me quickly implement this one.

ycombinator replied:
Hahaha, seems like I forgot about it too 😄. Thanks for finding it again — just re-request my review on the PR as usual when you have this implemented. Thanks!

mtojek replied:
Ok, now it's fixed :D

ycombinator commented on internal/testrunner/runners/system/servicedeployer/terraform.go:
```diff
@@ -0,0 +1,89 @@
+		return nil, errors.Wrap(err, "could not boot up service using docker compose")
+	}
+
+	// Build service container name
```
If I'm following correctly, we're creating a container for the terraform executor here, not the service being tested, right? If so, could you please fix the comment and variable names?

mtojek replied:
My bad, fixed. Thanks!

ycombinator commented on internal/testrunner/runners/system/servicedeployer/terraform.go:
```diff
@@ -0,0 +1,89 @@
+	serviceContainer := fmt.Sprintf("%s_terraform_1", service.project)
+	outCtxt.Hostname = serviceContainer
+
+	// Connect service network with stack network (for the purpose of metrics collection)
```
Apologies if you explained this to me already a couple days ago, but what is the reason for connecting the terraform executor to the stack network? Will any of the stack containers (e.g. agent) need to talk to the terraform executor container?

mtojek replied:
It was for next use cases, where we may want to proxy traffic from an external endpoint, but as long as I removed ports I think we can remove it too.

I think we'll have to check the progress of research on running the Elastic Agent in the k8s cluster and see how we can combine it with elastic-package.

ycombinator commented on internal/testrunner/runners/system/servicedeployer/terraform_env.go:
```diff
@@ -0,0 +1,53 @@
+	awsAccessKeyID     = "AWS_ACCESS_KEY_ID"
+	awsSecretAccessKey = "AWS_SECRET_ACCESS_KEY"
+	awsProfile         = "AWS_PROFILE"
+	awsRegion          = "AWS_REGION"
```
It would be really nice if we could keep any mention of integration-specific variables/constants out of the elastic-package source code. I know there is [another place](https://github.com/elastic/elastic-package/pull/227/files#r565746042) where this is happening as well. Similar to the other place, could you please add a TODO comment here with a link to an issue for cleaning this up?

mtojek replied:
https://github.com/elastic/elastic-package/issues/235

ycombinator commented on scripts/test-check-packages.sh:
```diff
@@ -5,6 +5,9 @@ set -euxo pipefail
 
+  # Dump stack logs
+  elastic-package stack dump -v --output build/elastic-stack-dump/check
+
```
Is this something you added just for debugging this PR or do you want to keep it?

mtojek replied:
I added this on purpose because during in the tests the stack is booted twice, but we dumped logs once. I needed this "dump" call to verify if AWS secrets are not leaking.

ycombinator reviewed (changes requested):
This is looking good. I left mostly minor comments. 

The big thing for me that needs to be addressed as part of this PR is updating the system test runner HOWTO documentation. Currently it talks only about the Docker Compose deployer. Please update it to talk about the Terraform deployer as well so users know what choices they have.

mtojek commented:
@ycombinator Thanks a lot for the review! I addressed your comments and introduced small changes in README. Please take a look and re-review whenever you've time.

ycombinator commented on docs/howto/system_testing.md:
```diff
@@ -39,19 +38,33 @@ Packages have a specific folder structure (only relevant parts shown).
+
+### Docker Compose service deployer
 
 The `docker-compose.yml` file defines the integration service(s) for the package. If your package has a logs data stream, the log files from your package's integration service must be written to a volume. For example, the `apache` package has the following definition in it's integration service's `docker-compose.yml` file.
```
This sentence made sense earlier because the sample file structure above showed a `docker-compose.yml` file. Now the sample file structure above just says `<service deployer files>` so this sentence doesn't make as much sense.

I would add another introductory sentence to bridge this gap, something like:

> When using the Docker Compose service deployer, the `<service deployer files>` must include a `docker-compose.yml` file.

mtojek replied:
Added here and for the TF service deployer

ycombinator commented on docs/howto/system_testing.md:
````diff
@@ -39,19 +38,33 @@ Packages have a specific folder structure (only relevant parts shown).
+            <service deployer files>
+```
+
+`<service deployer>` - a name of the supported service deployer: `docker` or `tf` (terraform).
````
Might be worth clarifying that the `docker` name is for the Docker Compose service deployer and the `tf` name is for the Terraform service deployer. It's pretty obvious but it doesn't hurt to be explicit and tie these names to the sections that follow.

mtojek replied:
Good point, docker and docker compose are different products.

ycombinator commented on docs/howto/system_testing.md:
```diff
@@ -66,7 +79,42 @@ services:
-### Data stream-level configuration
+### Terraform service deployer
+
+The `*.tf` files define the infrastructure using the Terraform syntax. The terraform based service can be handy to boot up
```
Similar to my comment on the Docker Composer service deployer, I would add an introductory sentence to bridge the gap between what's shown in the sample file structure and this sentence here.

mtojek replied:
Fixed

ycombinator commented on docs/howto/system_testing.md:
```diff
@@ -66,7 +79,42 @@ services:
+### Terraform service deployer
+
+The `*.tf` files define the infrastructure using the Terraform syntax. The terraform based service can be handy to boot up
+resources on AWS and use them for testing (e.g. spawn EC2 instance and collect its metrics).
```
Should we specifically mention AWS here? Or can we be more generic and say resources supported by Terraform or something like that?

mtojek replied:
Fixed

ycombinator commented on internal/testrunner/runners/system/servicedeployer/terraform.go:
```diff
@@ -0,0 +1,76 @@
+
+	// Build Terraform executor container name
+	serviceContainer := fmt.Sprintf("%s_terraform_1", service.project)
+	outCtxt.Hostname = serviceContainer
```
Do we need to set `outCtxt.Hostname` here? AFAICT it's not useful because the Agent configuration will never really need to talk to the Terraform executor container. Instead it will need to use the `outCtxt.CustomProperties` to talk to the service that is deployed by the Terraform executor.

mtojek replied:
We can try this, I hope everything will still work (it's not expected value somewhere). Fixed.

ycombinator commented on docs/howto/system_testing.md:
````diff
@@ -39,19 +38,33 @@ Packages have a specific folder structure (only relevant parts shown).
+        <service deployer files>
+```
+
+or the data-stream's level:
````
```suggestion
or the data stream's level:
```

ycombinator commented on docs/howto/system_testing.md:
```diff
@@ -39,19 +38,33 @@ Packages have a specific folder structure (only relevant parts shown).
-### Package-level configuration
-
-First, we must define the configuration for deploying a package's integration service. As mentioned in the [_Limitations_](#Limitations) section above, only packages whose integration services can be deployed via Docker Compose are supported at the moment.
+First, we must define the configuration for deploying a package's integration service. We can define it on either the package-level:
```
```suggestion
First, we must define the configuration for deploying a package's integration service. We can define it on either the package level:
```

ycombinator reviewed (approved):
LGTM.

ycombinator reviewed (approved):
LGTM.
